### PR TITLE
Feature/testing cost calculator

### PR DESCRIPTION
**Pull request by dragospopa**
JourneyCostCalculator tests done

### CODE REVIEW

File: src/com/tfl/billing/JourneyCostCalculator.java [hunk 1]
```diff
@@ -4,10 +4,7 @@
 import com.tfl.external.PaymentsSystem;
 
 import java.math.BigDecimal;
-import java.util.ArrayList;
-import java.util.Calendar;
-import java.util.Date;
-import java.util.List;
+import java.util.*;
```
Comment (seriousbee) [inline src/com/tfl/billing/JourneyCostCalculator.java]: Don't use * imports

File: src/com/tfl/billing/JourneyEvent.java [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 import java.util.UUID;
 
-public abstract class JourneyEvent {
```
Comment (seriousbee) [inline src/com/tfl/billing/JourneyEvent.java]: why is this no longer abstract? It actually makes sense for it to be abstact

File: src/com/tfl/billing/TravelTracker.java [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 import com.oyster.ScanListener;
 import com.tfl.external.Customer;
+import jdk.nashorn.internal.scripts.JO;
```
Comment (seriousbee) [inline src/com/tfl/billing/TravelTracker.java]: wtf is this?

File: test/com/tfl/billing/JourneyCostCalculatorTest.java [hunk 4]
```diff
@@ -2,70 +2,120 @@
...
 import org.jmock.integration.junit4.JUnitRuleMockery;
+import org.junit.Before;
 import org.junit.BeforeClass;
 import org.junit.Rule;
 import org.junit.Test;
 
 import java.math.BigDecimal;
-import java.util.Date;
-import java.util.List;
+import java.util.*;
```
Comment (seriousbee) [inline test/com/tfl/billing/JourneyCostCalculatorTest.java]: again, don't use * imports

File: src/com/tfl/billing/JourneyCostCalculator.java (hunk 1 above)
Comment (seriousbee) [reply src/com/tfl/billing/JourneyCostCalculator.java]: https://stackoverflow.com/questions/3348816/intellij-never-use-wildcard-imports

File: test/com/tfl/billing/JourneyCostCalculatorTest.java [hunk 5]
```diff
@@ -2,70 +2,120 @@
...
+    //initilaize required object
+    private List<JourneyEvent> eventLog = Arrays.asList(new JourneyEvent(c.cardId(), OysterReaderLocator.atStation(Station.PADDINGTON).id()), new JourneyEvent(c1.cardId(), OysterReaderLocator.atStation(Station.PADDINGTON).id()), new JourneyEvent(c.cardId(), OysterReaderLocator.atStation(Station.VICTORIA_STATION).id()), new JourneyEvent(c.cardId(), OysterReaderLocator.atStation(Station.OXFORD_CIRCUS).id()));
+    private List<Journey> customerJourneys = Arrays.asList(new Journey(new JourneyEvent(c.cardId(), OysterReaderLocator.atStation(Station.PADDINGTON).id()), new JourneyEvent(c.cardId(), OysterReaderLocator.atStation(Station.VICTORIA_STATION).id())));
+    JourneyCostCalculator costCalcuator = new JourneyCostCalculator(eventLog, customerJourneys);
+
 
-    @Rule
-    public JUnitRuleMockery context = new JUnitRuleMockery();
 
     private long getTimeInMillis(int hour) {
```
Comment (seriousbee) [inline test/com/tfl/billing/JourneyCostCalculatorTest.java]: consider renaming to `hoursToMillis`

File: test/com/tfl/billing/JourneyCostCalculatorTest.java [hunk 6]
```diff
@@ -2,70 +2,120 @@
...
 import java.math.BigDecimal;
-import java.util.Date;
-import java.util.List;
+import java.util.*;
 
 import static org.hamcrest.CoreMatchers.is;
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertThat;
-import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.*;
```
Comment (seriousbee) [inline test/com/tfl/billing/JourneyCostCalculatorTest.java]: again, pls no wildcards

File: test/com/tfl/billing/JourneyCostCalculatorTest.java [hunk 7]
```diff
@@ -2,70 +2,120 @@
...
     Date peakTime = new Date();
     Date offPeakTime = new Date();
     Customer c = CustomerDatabase.getInstance().getCustomers().get(0);
+    Customer c1 = CustomerDatabase.getInstance().getCustomers().get(1);
+    Customer c2 = CustomerDatabase.getInstance().getCustomers().get(2);
+
+    //initilaize required object
+    private List<JourneyEvent> eventLog = Arrays.asList(new JourneyEvent(c.cardId(), OysterReaderLocator.atStation(Station.PADDINGTON).id()), new JourneyEvent(c1.cardId(), OysterReaderLocator.atStation(Station.PADDINGTON).id()), new JourneyEvent(c.cardId(), OysterReaderLocator.atStation(Station.VICTORIA_STATION).id()), new JourneyEvent(c.cardId(), OysterReaderLocator.atStation(Station.OXFORD_CIRCUS).id()));
+    private List<Journey> customerJourneys = Arrays.asList(new Journey(new JourneyEvent(c.cardId(), OysterReaderLocator.atStation(Station.PADDINGTON).id()), new JourneyEvent(c.cardId(), OysterReaderLocator.atStation(Station.VICTORIA_STATION).id())));
+    JourneyCostCalculator costCalcuator = new JourneyCostCalculator(eventLog, customerJourneys);
```
Comment (seriousbee) [inline test/com/tfl/billing/JourneyCostCalculatorTest.java]: Why is all this stuff outside of any method?

File: test/com/tfl/billing/JourneyCostCalculatorTest.java [hunk 8]
```diff
@@ -2,70 +2,120 @@
...
 
     private long getTimeInMillis(int hour) {
         return hour*60*60*1000;
     }
 
     @Test
-    public void getJourneyEventsTest() throws Exception {
-        List<JourneyEvent> events = context.mock(List.class);
-
+    public void getAllCardScansForCustomer() throws Exception {
```
Comment (seriousbee) [inline test/com/tfl/billing/JourneyCostCalculatorTest.java]: why would this method throw exeptions? Maybe consider making another test which ensures that no exception is thrown? 
https://stackoverflow.com/questions/17731234/how-to-test-that-no-exception-is-thrown

File: test/com/tfl/billing/JourneyCostCalculatorTest.java [hunk 9]
```diff
@@ -2,70 +2,120 @@
...
     public void correctlyRoundstoTheNearestPenny(){
         System.out.println();
 
-        assertThat(testCalc.roundToNearestPenny(new BigDecimal(1.5110011)), is(new BigDecimal(1.51).setScale(2,BigDecimal.ROUND_HALF_UP)));
-        assertThat(testCalc.roundToNearestPenny(new BigDecimal(1.0100010)), is(new BigDecimal(1.01).setScale(2,BigDecimal.ROUND_HALF_UP)));
+        assertThat(costCalcuator.roundToNearestPenny(new BigDecimal(1.5110011)), is(new BigDecimal(1.51).setScale(2,BigDecimal.ROUND_HALF_UP)));
+        assertThat(costCalcuator.roundToNearestPenny(new BigDecimal(1.0100010)), is(new BigDecimal(1.01).setScale(2,BigDecimal.ROUND_HALF_UP)));
     }
 
 }
```
Comment (seriousbee) [inline test/com/tfl/billing/JourneyCostCalculatorTest.java]: add a newline at the end of the file

File: test/com/tfl/billing/JourneyCostCalculatorTest.java [hunk 10]
```diff
@@ -2,70 +2,120 @@
...
+
+    @Test
+    public void getTotalForNotTravellingCustomer() throws Exception {
+        eventLog = Arrays.asList(new JourneyStart(c.cardId(), OysterReaderLocator.atStation(Station.PADDINGTON).id()), new JourneyStart(c1.cardId(), OysterReaderLocator.atStation(Station.HOLBORN).id()), new JourneyEnd(c.cardId(), OysterReaderLocator.atStation(Station.VICTORIA_STATION).id()), new JourneyEnd(c1.cardId(), OysterReaderLocator.atStation(Station.CHANCERY_LANE).id()), new JourneyStart(c.cardId(), OysterReaderLocator.atStation(Station.OXFORD_CIRCUS).id()), new JourneyEnd(c.cardId(), OysterReaderLocator.atStation(Station.VICTORIA_STATION).id()));
+        BigDecimal expected = BigDecimal.ZERO;
+        assertThat(costCalcuator.roundToNearestPenny(costCalcuator.getTotalForCustomer(c2, eventLog)),is(costCalcuator.roundToNearestPenny(expected)));
     }
 
     @Test
     public void peakTest() {
```
Comment (seriousbee) [inline test/com/tfl/billing/JourneyCostCalculatorTest.java]: `peakTimeIsMarkedAsPeak`

**Review by seriousbee — CHANGES_REQUESTED** [review]
@dragospopa , pls review my comments

File: test/com/tfl/billing/JourneyCostCalculatorTest.java [hunk 11]
```diff
@@ -2,70 +2,120 @@
...
+            if(testResult.get(i).readerId()!=expected.get(i).readerId() || testResult.get(i).cardId()!=expected.get(i).cardId()){
+                flag = false;
+            }
+        }
+        assertTrue(flag);
     }
 
     @Test
-    public void getJourneysTest() throws Exception {
+    public void getAllCompletedJourneysForCustomerCurrentlyTravelling() throws Exception {
```
Comment (seriousbee) [inline test/com/tfl/billing/JourneyCostCalculatorTest.java]: this test is failing for me 🤔

File: test/com/tfl/billing/JourneyCostCalculatorTest.java [hunk 12]
```diff
@@ -2,70 +2,120 @@
...
+                    testResult.get(i).originId()!=expected.get(i).originId() ||
+                    testResult.get(i).destinationId()!=expected.get(i).destinationId()){
+                flag = false;
+            }
+        }
+        assertTrue(flag);
     }
 
     @Test
     public void getTotalTest() throws Exception {
```
Comment (seriousbee) [inline test/com/tfl/billing/JourneyCostCalculatorTest.java]: this test is failing for me 🤔

File: src/com/tfl/billing/JourneyCostCalculator.java (hunk 1 above)
Comment (dragospopa) [reply src/com/tfl/billing/JourneyCostCalculator.java]: my IDE is somehow importing stuff on it's own

File: src/com/tfl/billing/TravelTracker.java (hunk 3 above)
Comment (dragospopa) [reply src/com/tfl/billing/TravelTracker.java]: my IDE is somehow importing stuff on it's own x2

File: test/com/tfl/billing/JourneyCostCalculatorTest.java (hunk 4 above)
Comment (dragospopa) [reply test/com/tfl/billing/JourneyCostCalculatorTest.java]: my IDE is somehow importing stuff on it's own x3

File: test/com/tfl/billing/JourneyCostCalculatorTest.java (hunk 6 above)
Comment (dragospopa) [reply test/com/tfl/billing/JourneyCostCalculatorTest.java]: my IDE is somehow importing stuff on it's own x4

File: test/com/tfl/billing/JourneyCostCalculatorTest.java (hunk 7 above)
Comment (dragospopa) [reply test/com/tfl/billing/JourneyCostCalculatorTest.java]: shouldnt these be global? cause im using them in all the test methods and i dont want to have to create these every time

File: test/com/tfl/billing/JourneyCostCalculatorTest.java (hunk 5 above)
Comment (dragospopa) [reply test/com/tfl/billing/JourneyCostCalculatorTest.java]: 👍

File: test/com/tfl/billing/JourneyCostCalculatorTest.java [hunk 13]
```diff
@@ -16,56 +18,105 @@
...
+        assertTrue(costCalcuator.peak(peakTime));
+        assertFalse(costCalcuator.peak(offPeakTime));
 
-        assertTrue(testCalc.peak(peakTime));
-        assertFalse(testCalc.peak(offPeakTime));
+        peakTime.setTime(hoursToMillis(18));
+        offPeakTime.setTime(hoursToMillis(22));
 
+        assertTrue(costCalcuator.peak(peakTime));
+        assertFalse(costCalcuator.peak(offPeakTime));
```
Comment (seriousbee) [inline test/com/tfl/billing/JourneyCostCalculatorTest.java]: having multiple assserts in a test is rather a bad practice. https://softwareengineering.stackexchange.com/questions/7823/is-it-ok-to-have-multiple-asserts-in-a-single-unit-test

File: test/com/tfl/billing/JourneyCostCalculatorTest.java [hunk 14]
```diff
@@ -16,56 +18,105 @@
...
 
     @Test
-    public void correctlyRoundstoTheNearestPenny(){
-        System.out.println();
-
-        assertThat(testCalc.roundToNearestPenny(new BigDecimal(1.5110011)), is(new BigDecimal(1.51).setScale(2,BigDecimal.ROUND_HALF_UP)));
-        assertThat(testCalc.roundToNearestPenny(new BigDecimal(1.0100010)), is(new BigDecimal(1.01).setScale(2,BigDecimal.ROUND_HALF_UP)));
+    public void correctlyRoundstoTheNearestPenny() {
+        assertThat(costCalcuator.roundToNearestPenny(new BigDecimal(1.5110011)), is(new BigDecimal(1.51).setScale(2,BigDecimal.ROUND_HALF_UP)));
+        assertThat(costCalcuator.roundToNearestPenny(new BigDecimal(1.0100010)), is(new BigDecimal(1.01).setScale(2,BigDecimal.ROUND_HALF_UP)));
```
Comment (seriousbee) [inline test/com/tfl/billing/JourneyCostCalculatorTest.java]: again, pls only one assert per test

File: test/com/tfl/billing/JourneyCostCalculatorTest.java [hunk 15]
```diff
@@ -16,56 +18,105 @@
...
     }
 
-}
```
Comment (seriousbee) [inline test/com/tfl/billing/JourneyCostCalculatorTest.java]: lovely <3

**Review by seriousbee — CHANGES_REQUESTED** [review]

File: test/com/tfl/billing/JourneyCostCalculatorTest.java (hunk 15 above)
Comment (dragospopa) [reply test/com/tfl/billing/JourneyCostCalculatorTest.java]: 🥇

**Review by seriousbee — APPROVED** [review]

**Pull request merged**